WillCWX/markbind
# Bruh

**What is the purpose of this pull request?**




**Overview of changes:**

Feature / Behaviour Changes:
- `markbind init --convert` 
  - On an empty folder now produces a website that is exactly the same as `markbind  init` 
    - This also means no more `about.md` as it does not appear in `template/default.md`
    - The icon is also changed into `Your Logo` as it is in `default.md`
  - No longer adds `404` as a navigation link, `404.html` is also now displayed exactly the same as in `markbind init`
  - Is now able to be used with `-t minimal` and have `default.md` be exactly the same as a minimal template
  - generated `site.json` is now more readable with indentations
  - `footer` and `siteNav` nunjucks variables no longer replace their defaults and instead adds to their defaults.
- `markbind init`
  - `default.md` is now generated by `Site.generateDefaultLayout` rather than copied over

Code Maintenance Changes:

- Updating `default.md` layout:
  - As `default.md` should no longer exist, update the respective template `.njk` nunjucks files instead.
  -  New `TemplateConfig.ts` contains template configurations for both default and minimal.
       -  `njkFile: string` dictates which nunjucks file to use.
       -  `njkSubs: NjkVars[]` allows for using nunjucks to copy existing files to incorporate into the layout `default.md`
       -  `layoutObjs: SiteConfigPage[]` allows for edits to `config`
       -  `hasAutoSiteNav: boolean` toggles whether `Site.buildSiteNav` is called during a conversion
       -  `siteNavIgnore: string[]` contains files that `Site.buildSiteNav` will ignore when making links 



**Anything you'd like to highlight/discuss:**
I have no idea whether this was worth it.
The scope of `markbind init --convert` is broad and its functions could potentially be used for easy migrations / upgrades.
However, this increase in scope may also unduly burden the `Site` class as it is the only class currently able to inherently do everything needed to convert. 

I'll add more tests and update the docs once we discuss whether this was worth it...
I have not deleted the `default.md` in the `core/template/...` for easier cross reference.

**Testing instructions:**
`npm run test` passes. The functional tests for `test_site_templates` have not changed but should still pass because the generated default.md for both minimal and default is exactly the same.


Fast-ish Manual Tests:
  1. Copy the `non-markbind-site` from the conversion tests
  2. `npm link` with this branch
  3. `markbind init --convert` the copied `non-markbind-site` folders
  4. `markbind serve` to see that they look `markbind init` but with texts different in the front page and additional / inherited links at the side. Additional texts to footer should exist if `_Footer.md` is present.



**Proposed commit message: (wrap lines at 72 characters)**
Update Site to generate default.md given template



---

**Checklist:** :ballot_box_with_check:
